pkg/apis/workspaces/v1alpha1/projects_conversion.go

```diff
@@ -19,8 +19,6 @@ func convertProjectTo_v1alpha2(src *Project, dest *v1alpha2.Project) error {
 	switch {
 	case src.Git != nil:
 		sparseCheckoutDir = src.Git.SparseCheckoutDir
-	case src.Github != nil:
-		sparseCheckoutDir = src.Github.SparseCheckoutDir
```

Comment:
since `Github` is duplicate of `Git`. Should we convert `v1.Github` to `v2.Git`, instead of ignoring it?

Reply:
Hm, this might be a good idea. My first impulse is to not do it since it would change the file on round trip (i.e. `v1alpha1 (GitHub) -> v1alpha2 (Git) -> v1alpha1 (Git)`), but my approach of ignoring the problem just changes the file in a different way. 

On the other hand, if there's not a whole lot of value in it, it's a lot simpler to not bother. The way conversion is implemented (round trip through json) would mean we'd have to do something fancy to make it work.

Reply:
emm. When you say "round trip conversion", did you mean how the conversion test is implemented?
I agree more extra work needs to be done if we want to test github->git conversion. however, the `ConvertTo` (https://github.com/devfile/api/blob/a313872749ed60b47fb015288bb6595df64dac33/pkg/apis/workspaces/v1alpha1/devworkspace_conversion.go#L11) function is exposed, and can be called by consumer. so we probably should not simply ignore it. 
what if we still convert `v1alpha1 (GitHub) -> v1alpha2 (Git)` in the actual conversion function, and just ignore `Github` case in the conversion test?

Reply:
The way conversion is supposed to be implemented (from a k8s perspective) is that converting from version X to the hub version Y (v1alpha2 in our case) and back again should result in unmodified data. This is because internally, when you apply a v1alpha1 DevWorkspace to your cluster, kube will internally convert it to v1alpha2 and store that in etcd. This v1alpha2 spec is converted back to v1alpha1 when you request it from the cluster -- this is what the conversion tests are testing.

The way this is often done for removed/added fields is to serialize removed fields into a json annotation so that when you unconvert you get the same data, but IMO this is a very messy approach for a feature nobody is using (i.e. a bunch of work for no gain). We've done a similar thing for e.g. the vscode command types.

The additional work that would be required to implement GitHub -> Git conversion is due to the approach we take to conversion, where we take version X, serialize it to json, and then deserialize it to version Y (e.g. [here](https://github.com/devfile/api/blob/cd9c30e6aa05b15445bb05386692f470323c826f/pkg/apis/workspaces/v1alpha1/projects_conversion.go#L10-L17)). This was done since the main breaking change between v1alpha1 and v1alpha2 is moving component name, command id, etc. one level higher in the spec (i.e. `.spec.components[].name` vs `.spec.components[].container.name`, which makes the types not assignable and makes conversion-gen code nearly useless. By going through json, we can easily assign all fields shared between v1alpha1 and v1alpha2, but this doesn't work if the path is different (deserializing a GitHub project -- e.g. `.projects[].github.remotes` -- will deserialize to nil if you pass in a `Git` project). 

I can definitely look into it, but I'm hesitant to introduce complexity in the conversion code to cover a case that (I think) will never actually be called in production.

Reply:
I'm thinking if adding an attribute when convert Github -> Git would help in this situation?

Reply:
The basic gist of the approach is
```
        v1alpha1             |     intermediate (v1alpha1)    |            v1alpha2              
-----------------------------|--------------------------------|--------------------------------
projects:                    |   projects:                    |   projects:                     
  - name: project            |     - name: project            |     - name: project   
    github:                  |       github:                  |                                 
      remotes:               |         remotes:               |                                 
        origin: my-url.com   |           origin: my-url.com   |                                 
                             |       git:                     |       git:                      
                             |         remotes:               |         remotes:                
                             |           origin: my-url.com   |           origin: my-url.com    
```